bpi.syndicate.inc

```diff
@@ -543,3 +688,217 @@ function bpi_syndicate_action($bpi_id) {
+    $name = 'logicalOperator';
+    $id = $name;
+    $facets_form[$id] = array(
+      '#type' => 'select',
+      '#title' => bpi_label_mapper($name) ? bpi_label_mapper($name) : $name,
+      '#options' => array(
+        'and' => t('All must apply'),
+        'or' => t('Any must apply'),
+      ),
+      '#default_value' => _bpi_get_query_parameter(array('filter', $name), 'and'),
+      '#weight' => -10,
+    );
+
+    $filter_form['bpi_facets'] = $facets_form;
+
+    $filter_form['bpi_actions'] = array(
+      '#type' => 'container',
+      '#attributes' => array(
+        'class' => array(
+          drupal_html_class('container-inline'),
+        ),
+      ),
+    );
+
+    if (_bpi_get_current_filter()) {
+      $filter_form['bpi_actions']['reset_filter'] = array(
+        '#type' => 'item',
+        '#markup' => l(t('Reset filters'), current_path(), array(
+          'query' => _bpi_get_current_search_query(),
+          'attributes' => array(
+            'class' => array('button'),
+          ),
+        )),
+      );
+    }
+
+    $form['bpi_filter'] = $filter_form;
+  }
+}
+
+/**
+ * Get facets as an array.
+ *
+ * For performance reasons facets will only be loaded when no search or
+ * filters are applied.
+ *
+ * @return array
+ *   All facets.
+ */
+function bpi_get_facets() {
+  // Reload facets if they're not cached locally.
+  $reload = !isset($_SESSION['bpi_all_facets']) || !is_array($_SESSION['bpi_all_facets']);
+  if (!$reload) {
+    $query = _bpi_get_query_parameter();
+    // Reload if neither text nor filter is applied.
+    $reload = !_bpi_get_current_search_phrase() || !_bpi_get_current_filter();
+  }
+
+  if ($reload) {
+    $response = bpi_search_content();
+    if ($response) {
+      $facets = $response->getFacets();
+      $_SESSION['bpi_all_facets'] = bpi_facets_to_array($facets);
+    }
+    else {
+      $_SESSION['bpi_all_facets'] = NULL;
+    }
+  }
+
+  return $_SESSION['bpi_all_facets'];
+}
+
+/**
+ * Set search facets.
+ *
+ * Used to store facets from actual search in the session as we need the
+ * facets when rendering the form on postback.
+ *
+ * @param \Bpi\Sdk\Facets $facets
+ *   The facets returned from the search.
+ */
+function bpi_set_search_facets(\Bpi\Sdk\Facets $facets = NULL) {
+  // Convert facets to array for safe storage in session.
+  $_SESSION['bpi_search_facets'] = $facets ? bpi_facets_to_array($facets) : NULL;
+}
+
+/**
+ * Get facets from last search.
+ *
+ * @return array $facets
+ *   The facets.
+ */
+function bpi_get_search_facets() {
+  return isset($_SESSION['bpi_search_facets']) ? $_SESSION['bpi_search_facets'] : NULL;
+}
+
+/**
+ * Convert facets into an array.
+ *
+ * @param \Bpi\Sdk\Facets $facets
+ *   The facets.
+ *
+ * @return array
+ *   The facets as an associative array.
+ *   term => [
+ *     title => term title,
+ *     count => item count
+ *   ]
+ */
+function bpi_facets_to_array(\Bpi\Sdk\Facets $facets) {
+  $data = NULL;
+
+  if ($facets) {
```

Comment:
What is the default value for facets? The parameter requires it to be \Bpi\Sdk\Facets so would this not always be true